### DISCUSSION TITLE
Incorrect work with ldap (active directory)

### DISCUSSION THREAD

**Discussion by maximmm92** [ClickHouse/ClickHouse#53507] [Q&A] [answered]
I have an active directory, there are groups and users in it, if I want to connect to clickhouse using ldap and the roles in clickhouse are the same as the groups in ad, then everything works fine, but if the user is placed in a group that does not have a role in clickhouse, and then remove it from there, then when connecting to ckuckhouse he starts to wait for CLOSE_WAIT, what is the reason, is this bug or can it be fixed somehow?

`<my_ad_server>
            <host>ХХ.ХХ.ХХХ.ХХХ</host>
            <port>389</port>
            <verification_cooldown>100</verification_cooldown>
            <bind_dn>SOMECH\{user_name}</bind_dn>
            <user_dn_detection>
                <base_dn>OU=CHILD,DC=SOMECH,DC=XXX,DC=yyy</base_dn>
                <search_filter>(&amp;(objectClass=user)(sAMAccountName={user_name}))</search_filter>
            </user_dn_detection>
            <enable_tls>no</enable_tls>
</my_ad_server>`

`<ldap>
            <server>my_ad_server</server>
            <role_mapping>
                <base_dn>OU=CHILD,DC=SOMECH,DC=XXX,DC=yyy</base_dn>
                <attribute>CN</attribute>
                <scope>subtree</scope>
                <search_filter>(&amp;(objectClass=group)(member={user_dn}))</search_filter>
            </role_mapping>
</ldap>`

**Comment by tsolodov** [comment]
I do not have AD to test this setup, but I would check network connectivity `nc -vz ldap_server 389`. If you have connectivity, try to query AD using third-party tools(ldapsearch/ Apache Directory Studio). I recommend to follow official documentation: https://clickhouse.com/docs/en/guides/sre/configuring-ldap#1-configure-ldap-connection-settings-in-clickhouse

**Reply by maximmm92** [reply]
Сolleague, thank you for your answer, but there are no problems with the network and ad is configured according to the documentation.

**Comment by mark-bb** [comment]
We opened an issue on this:
https://github.com/ClickHouse/ClickHouse/issues/53708
The problem is reproduced easily on all recent CH versions and platforms.

Briefly: it's almost impossible to use LDAP authentication and role mapping (against MSAD at least) in a large organization (with many users and groups) with this bug.

**Reply by tsolodov** [reply]
Checked the issue.

You can use `<prefix>clickhouse_role_</prefix>` in role_mapping section to filter groups that supposed to be ClickHouse roles. 

According to docs, each role expected to be created beforehand by the CREATE ROLE statement, so if you add new role(group membership in AD), make sure the role is created beforehand.

**Reply by mark-bb** [reply]
Sure, it's one of the workarounds - to create such prefixed groups for each current user group, include each current user group to the corresponding prefixed one, and instruct LDAP admins to not touch these prefixed groups without informing a CH admin.

But it's impractical in large organizations - there is a lot of users and groups in LDAP, and it's hard to LDAP admins to ask some product admin if it's possible to add / remove some particular user to / from some particular group every time.

"expected to be created" and "has to / must be created" are different suggestions. So, I believe that it's a bug.

**Reply by tsolodov** [reply]
Hi @mark-bb, I think there are 2 separate issues:
- ClickHouse should handle correctly misconfiguration when profile is not configured and  if [bug](https://github.com/ClickHouse/ClickHouse/issues/53708) is confirmed, it will be addressed.
- How to structure resources(like security groups, organization uints, Users) in Active Directory(or LDAP) and how to handle changes. This is really big topic and there is no right solution. Each company has own procedures and protocols.

**Reply by mark-bb** [reply]
Hi @tsolodov 
It seems that the problem was resolved, but my last question in the issue has not been answered.
Is it possible to understand which fixpack of every lts release (23.3, 23.8, etc.) will include this resolution?
I see some info on cherry picks & backports in [#55119](https://github.com/ClickHouse/ClickHouse/pull/55119) related to the issue, but I can't interpret it properly...

**Reply by tsolodov** [reply]
Hi @mark-bb , so far it's part of 3 releases:
- https://github.com/ClickHouse/ClickHouse/releases/tag/v23.9.2.56-stable
- https://github.com/ClickHouse/ClickHouse/releases/tag/v23.8.4.69-lts
- https://github.com/ClickHouse/ClickHouse/releases/tag/v23.3.14.78-lts

**Comment by mark-bb** [comment]
Hi @tsolodov,

According to the following links with comparison of the latest versions with the corresponding previous ones (looking at the presence of #55119 there) the patch was included into 23.9.2, 23.8.4, but not into 23.3.14.

https://github.com/ClickHouse/ClickHouse/compare/v23.9.1.1854-stable...v23.9.2.56-stable
https://github.com/ClickHouse/ClickHouse/compare/v23.8.3.48-lts...v23.8.4.69-lts
https://github.com/ClickHouse/ClickHouse/compare/v23.3.13.6-lts...v23.3.14.78-lts

According to the link below the problem is still in the "Open" status for 23.3 as well.
https://github.com/ClickHouse/ClickHouse/pull/55119

So, the question is: are the any plans to include the patch into some 23.3 fix?

**Reply by tsolodov** [reply]
Hi @mark-bb, there was a merge conflict in PR, now it's resolved and should be merged to 23.3 soon.

**Reply by mark-bb** [reply]
Hi @tsolodov, 
It seems, that this patch is still not included into the latest 23.3 build - 23.3.18.15-lts
https://github.com/ClickHouse/ClickHouse/compare/v23.3.13.6-lts...v23.3.18.15-lts

Am I correct?
If yes, then are there any plans to include the patch into 23.3?
If no, then is there some better way then link above to understand if this patch is really included into some 23.3 build?

**Reply by alexey-milovidov** [reply, marked as answer]
There were complications during backport, and we decided not to proceed with it.
You can upgrade to the version 23.12.
